server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeMetaWriter.java

```diff
@@ -0,0 +1,226 @@
+        int headerSizeInBytes = 16;
+
+        // For star-tree field name
+        headerSizeInBytes += starTreeFieldName.getBytes(UTF_8).length;
+
+        // For star tree field type
+        headerSizeInBytes += compositeFieldType.getName().getBytes(UTF_8).length;
+
+        return headerSizeInBytes;
+    }
+
+    /**
+     * Writes the star-tree metadata header.
+     *
+     * @param metaOut            the IndexOutput to write the header
+     * @param compositeFieldType the composite field type of the star-tree field
+     * @param starTreeFieldName  the name of the star-tree field
+     * @throws IOException if an I/O error occurs while writing the header
+     */
+    private static void writeMetaHeader(
+        IndexOutput metaOut,
+        CompositeMappedFieldType.CompositeFieldType compositeFieldType,
+        String starTreeFieldName
+    ) throws IOException {
+        // magic marker for sanity
+        metaOut.writeLong(MAGIC_MARKER);
+
+        // version
+        metaOut.writeInt(VERSION);
+
+        // star tree field name
+        metaOut.writeString(starTreeFieldName);
+
+        // star tree field type
+        metaOut.writeString(compositeFieldType.getName());
+    }
+
+    /**
+     * Writes the star-tree metadata.
+     *
+     * @param metaOut                   the IndexOutput to write the metadata
+     * @param writeState                the segment write state
+     * @param metricAggregatorInfos     the list of metric aggregator information
+     * @param starTreeField             the star tree field
+     * @param segmentAggregatedDocCount the aggregated document count for the segment
+     * @param dataFilePointer           the file pointer to the start of the star-tree data
+     * @param dataFileLength            the length of the star-tree data file
+     * @throws IOException if an I/O error occurs while writing the metadata
+     */
+    private static void writeMeta(
+        IndexOutput metaOut,
+        SegmentWriteState writeState,
+        List<MetricAggregatorInfo> metricAggregatorInfos,
+        StarTreeField starTreeField,
+        Integer segmentAggregatedDocCount,
+        long dataFilePointer,
+        long dataFileLength
+    ) throws IOException {
+
+        // number of dimensions
+        metaOut.writeInt(starTreeField.getDimensionsOrder().size());
+
+        // dimensions
+        for (Dimension dimension : starTreeField.getDimensionsOrder()) {
+            int dimensionFieldNumber = writeState.fieldInfos.fieldInfo(dimension.getField()).getFieldNumber();
+            metaOut.writeInt(dimensionFieldNumber);
+        }
+
+        // number of metrics
+        metaOut.writeInt(metricAggregatorInfos.size());
+
+        // metric - metric stat pair
+        for (MetricAggregatorInfo metricAggregatorInfo : metricAggregatorInfos) {
+            String metricName = metricAggregatorInfo.getField();
+            String metricStatName = metricAggregatorInfo.getMetricStat().getTypeName();
```

Comment:
same comment as build mode enum, can't we use int constants to identify each unique metric.